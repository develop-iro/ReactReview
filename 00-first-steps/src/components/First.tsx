import ItemCounter from "./ItemCounter";
// In React, <> and </> are called fragments. They let you group multiple elements without adding extra nodes to the DOM.

// Purpose:
// A fragment allows you to return several elements from a component without wrapping them in a parent <div> or other HTML tag.

// Example:
const listOfCounters = [1, 2, 3];

function First() {
  return (
    <>
      <h1>My First Component</h1>
      <p>This is a simple component.</p>

      {listOfCounters.map((counter) => (
        <ItemCounter key={counter} initialCount={counter} />
      ))}
    </>
  );
}

export default First;
