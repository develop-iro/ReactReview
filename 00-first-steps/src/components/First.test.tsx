// This test file is intended for testing the First component using Vitest and React Testing Library.
//
// Concepts to explain and possible mocks:
// - Mocking props: You can pass different props to the component to test various scenarios.
// - Mocking child components: Use vi.mock or manual mocks to replace child components with simple versions.
// - Mocking functions: If your component receives callback props, you can use jest.fn() or vi.fn() to track calls and arguments.
// - Mocking modules: Use vi.mock() to mock imports like helpers or API calls.
// - Mocking context: If your component uses React context, you can wrap it with a custom provider in your tests.
// - Snapshot: Use toMatchSnapshot to ensure the rendered output remains consistent over time.
//
// Example mocks you can use in this suite:
// - Mock ItemCounter to isolate First's logic:
//   vi.mock('./ItemCounter', () => ({ default: () => <div>Mocked Counter</div> }));
// - Mock a helper function:
//   vi.mock('../helpers/format.helper', () => ({ formatCurrency: vi.fn() }));
// - Mock a callback prop:
//   const mockCallback = vi.fn();
//   render(<First onAction={mockCallback} />);
//
// These mocks help you isolate the component under test and control its dependencies, making your tests more reliable and focused.
import { describe, test, expect, vi, afterEach } from "vitest";
import { render } from "@testing-library/react";
import First from "./First";
vi.mock("./ItemCounter", () => ({
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  default: (props: unknown) => <div>Mocked Counter</div>,
}));

describe("First Component", () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  test("should match snapshot", () => {
    const { container } = render(<First />);
    expect(container).toMatchSnapshot();
  });
  test("should render heading and paragraph", () => {
    const { getByText } = render(<First />);
    expect(getByText("My First Component")).toBeTruthy();
    expect(getByText("This is a simple component.")).toBeTruthy();
  });
  test("should render mocked ItemCounter components", () => {
    const { getAllByText } = render(<First />);
    const counters = getAllByText("Mocked Counter");
    expect(counters.length).toBe(3);
  });
});
