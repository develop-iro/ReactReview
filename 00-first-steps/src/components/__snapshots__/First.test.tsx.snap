// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`First Component > should match snapshot 1`] = `
<div>
  <h1>
    My First Component
  </h1>
  <p>
    This is a simple component.
  </p>
  <div>
    Mocked Counter
  </div>
  <div>
    Mocked Counter
  </div>
  <div>
    Mocked Counter
  </div>
</div>
`;
